File: rtsp_video_downloader.cpp
#include <opencv2/opencv.hpp>
#include <boost/program_options.hpp>
#include <boost/filesystem.hpp>
#include <boost/date_time/posix_time/posix_time.hpp>
#include <iostream>
#include <vector>
#include <thread>
#include <chrono>

namespace po = boost::program_options;
namespace fs = boost::filesystem;

void rtsp_record(std::string cam_username, std::string cam_pass, std::string ip, int num, int fps, bool save_video, bool live, std::string start_time, int end_time, std::string out_dir) {
    std::string RTSP_URL;
    if (live) {
        RTSP_URL = "rtsp://" + cam_username + ":" + cam_pass + "@" + ip + "/ISAPI/Streaming/channels/" + std::to_string(num) + "01";
    } else {
        RTSP_URL = "rtsp://" + cam_username + ":" + cam_pass + "@" + ip + "/ISAPI/Streaming/tracks/" + std::to_string(num) + "01?starttime=" + start_time + "Z";
    }

    std::cout << "RTSP URL: " << RTSP_URL << std::endl;

    cv::VideoCapture cap(RTSP_URL, cv::CAP_FFMPEG);
    if (!cap.isOpened()) {
        std::cerr << "Cannot open RTSP stream." << std::endl;
        return;
    }

    int frame_width = static_cast<int>(cap.get(cv::CAP_PROP_FRAME_WIDTH));
    int frame_height = static_cast<int>(cap.get(cv::CAP_PROP_FRAME_HEIGHT));
    std::string to_date = boost::posix_time::to_iso_string(boost::posix_time::second_clock::local_time());
    std::string full_out_dir = fs::path(out_dir).append(to_date).string();
    fs::create_directories(full_out_dir);

    std::string video_filename = to_date + "_cam_" + std::to_string(num) + ".mp4";
    std::string video_filepath = fs::path(full_out_dir).append(video_filename).string();
    cv::VideoWriter video_output(video_filepath, cv::VideoWriter::fourcc('X', 'V', 'I', 'D'), fps, cv::Size(frame_width, frame_height), true);

    std::cout << "[INFO]: RTSP Streaming Started....." << std::endl;
    auto start = std::chrono::system_clock::now();
    int frame_count = 0;

    while (true) {
        cv::Mat frame;
        if (!cap.read(frame)) {
            break;
        }
        frame_count++;
        auto now = std::chrono::system_clock::now();
        auto elapsed_time = std::chrono::duration_cast<std::chrono::seconds>(now - start).count();

        if (save_video && (elapsed_time < end_time * 60)) {
            video_output.write(frame);
            std::cout << "[INFO]: Saving video frame " << frame_count << " at " << fps << " fps....cam " << num << std::endl;
        }

        if (cv::waitKey(1) == 'q' || elapsed_time >= end_time * 60) {
            std::cout << "\n [INFO]: Stopping recording..." << std::endl;
            break;
        }
    }

    cap.release();
    video_output.release();
    std::cout << "All videos saved successfully.!" << std::endl;
}

int main(int argc, char* argv[]) {
    po::options_description desc("Allowed options");
    desc.add_options()
        ("cam_username", po::value<std::string>()->default_value("admin"), "ip_camera's username")
        ("cam_pass", po::value<std::string>()->default_value("admin"), "ip_camera's password")
        ("ip", po::value<std::string>()->default_value("192.168.0.5"), "ip_camera's ip address")
        ("cam_list", po::value<std::vector<int>>()->multitoken()->required(), "Set camera range")
        ("save_video", po::value<bool>()->default_value(true), "Save the RTSP video")
        ("fps", po::value<int>()->default_value(4), "FPS")
        ("live", po::value<bool>()->default_value(false), "Live record or Tracks playback record")
        ("start_time", po::value<std::string>()->default_value("20240423T114200"), "Playback start-time")
        ("end_time", po::value<int>()->default_value(5), "Playback end-time in minutes")
        ("out_dir", po::value<std::string>()->default_value(boost::posix_time::to_iso_string(boost::posix_time::second_clock::local_time())), "Output directory");

    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);

    std::vector<int> cam_list = vm["cam_list"].as<std::vector<int>>();
    std::vector<std::thread> threads;

    for (int i = cam_list[0]; i <= cam_list[1]; ++i) {
        threads.emplace_back(rtsp_record, vm["cam_username"].as<std::string>(), vm["cam_pass"].as<std::string>(), vm["ip"].as<std::string>(), i, vm["fps"].as<int>(), vm["save_video"].as<bool>(), vm["live"].as<bool>(), vm["start_time"].as<std::string>(), vm["end_time"].as<int>(), vm["out_dir"].as<std::string>());
    }

    for (auto& thread : threads) {
        thread.join();
    }

    return 0;
}

File: rtsp_video_downloader.py
import os
import cv2
import argparse
import threading
from datetime import datetime, timedelta


def get_parser():
    arg_parser = argparse.ArgumentParser()
    arg_parser.add_argument('--cam_username', action='store', type=str, required=False, default='admin', help="ip_camera's username")
    arg_parser.add_argument('--cam_pass', action='store', type=str, required=False, default='admin', help="ip_camera's password")
    arg_parser.add_argument('--ip', action='store', type=str, required=False, default='192.168.0.1', help="ip_camera's ip address")
    arg_parser.add_argument('--cam_list', nargs='+', help='<Required> Set flag', type=int,required=True)
    arg_parser.add_argument('--save_video', action='store', type=bool, required=False, default=True , help="Save the RTSP video in recordings directory")
    arg_parser.add_argument('--fps', action='store', type=int, required=False, default=4, help="FPS")
    arg_parser.add_argument('--live', action='store', type=bool, required=False, default=False , help="Live record or Tracks playback record")
    arg_parser.add_argument('--start-time', action='store', type=str, required=False, default='20240423T114200', help="playback_start-time")
    arg_parser.add_argument('--end-time', action='store', type=int, required=False, default=5, help="playback_end-time in minutes")
    arg_parser.add_argument('--out_dir', action='store', type=str, required=False, default=datetime.now().strftime("%Y-%m-%d"), help="save_video_dir")

    return arg_parser.parse_args()


def rtsp_record(cam_username, cam_pass, ip, num, desired_fps, save_video, live, start_time, end_time, out_dir):
    # Construct RTSP URL based on whether it's live or playback
    if live:
        RTSP_URL = f'rtsp://{cam_username}:{cam_pass}@{ip}/ISAPI/Streaming/channels/{num}01'
    else:
        RTSP_URL = f'rtsp://{cam_username}:{cam_pass}@{ip}/ISAPI/Streaming/tracks/{num}01?starttime={start_time}Z'

    print(f"RTSP URL: {RTSP_URL}")

    # Set RTSP options
    os.environ['OPENCV_FFMPEG_CAPTURE_OPTIONS'] = 'rtsp_transport;tcp'
    #os.environ["OPENCV_FFMPEG_CAPTURE_OPTIONS"]="video_codec;h264_cuvid"

    # Start capturing the RTSP stream
    cap = cv2.VideoCapture(RTSP_URL, cv2.CAP_FFMPEG)
    if not cap.isOpened():
        print('Cannot open RTSP stream.')
        return

    # Create output directory structure
    to_date = datetime.now().strftime("%Y_%m_%d")
    full_out_dir = os.path.join(out_dir, to_date)
    os.makedirs(full_out_dir, exist_ok=True)

    # Setup video writer
    frame_width = int(cap.get(3))
    frame_height = int(cap.get(4))
    
    video_filename = f"{datetime.now().strftime('%Y_%m_%d-%I_%M_%S_%p')}_cam_{num}.mp4"
    video_filepath = os.path.join(full_out_dir, video_filename)
    video_codec = cv2.VideoWriter_fourcc(*'XVID')
    video_output = cv2.VideoWriter(video_filepath, video_codec, desired_fps, (frame_width, frame_height))

    print("[INFO]: RTSP Streaming Started.....")

    # Initialize frame timing and counting
    start = datetime.now()
    frame_count = 0

    while True:
        ret, frame = cap.read()
        if ret:
            frame_count += 1
            now = datetime.now()
            elapsed_time = (now - start).total_seconds()

            actual_fps = frame_count / elapsed_time if elapsed_time > 0 else 0
            frame_skip_ratio = actual_fps / desired_fps if desired_fps > 0 else 1

            if save_video and frame_skip_ratio > 0 and frame_count % int(max(1, frame_skip_ratio)) == 0:
                video_output.write(frame)
                print(f"[INFO]: Saving video frame {frame_count} at {desired_fps} fps....cam {num}")

            if cv2.waitKey(1) & 0xFF == ord('q') or (elapsed_time >= end_time*60):    # Stop recording after end_time minutes = 5 minutes by default
                print("\n [INFO]: Stopping recording...")
                break
        else:
            break

    cap.release()
    video_output.release()
    cv2.destroyAllWindows()
    print('All videos saved successfully.!')


if __name__ =='__main__':
    args = get_parser()

    camera_list =list(range(args.cam_list[0] ,args.cam_list[1]+1))
    print("List of cameras for recording...", camera_list)
    thread_list = []

    for i in range(len(camera_list)):
        thread = threading.Thread(target=rtsp_record,args=(args.cam_username, args.cam_pass, args.ip, camera_list[i], args.fps, args.save_video, args.live, args.start_time, args.end_time, args.out_dir),daemon=False)
        thread_list.append(thread)

    for thread in thread_list:
        thread.start()
